tensorboard/db.py

```diff
@@ -20,160 +20,492 @@
+    be fetched at once.
+
+    Fields:
+      rowid: The rowid which has an arbitrary number in the low 38 bits
+          and the experiment ID in the higher 28 bits. This is used to
+          control locality.
+      tag_id: The ID of the related row in the Tags table.
+      plugin_id: The ID of the related row in the Plugins table.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS TagsPlugins (
+          rowid INTEGER PRIMARY KEY,
+          tag_id INTEGER NOT NULL,
+          plugin_id INTEGER NOT NULL
+        )
+      ''')
+
+  def create_tensors_table(self):
+    """Creates the Tensors table.
+
+    This table stores the individual tensors.
+
+    Fields:
+      rowid: A 63-bit number containing the step count in the low 32
+          bits, and the randomly generated tag ID in the higher 31 bits.
+      encoding: A number indicating how the tensor was encoded to the
+          tensor blob field. 0 indicates an uncompressed binary Tensor
+          proto. 1..9 indicates a binary Tensor proto gzipped at the
+          corresponding level. 10..255 are reserved for future encoding
+          methods.
+      is_big: A boolean indicating that the tensor field is empty and a
+          separate asynchronous lookup should be performed on the
+          BigTensors table.
+      tensor: A binary representation of this tensor. This will be empty
+          if the is_big field is true.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS Tensors (
+          rowid INTEGER PRIMARY KEY,
+          encoding TINYINT NOT NULL,
+          is_big BOOLEAN NOT NULL,
+          tensor BLOB NOT NULL  -- VARBINARY on MySQL, MS-SQL, etc.
+        )
+      ''')
+
+  def create_big_tensors_table(self):
+    """Creates the BigTensors table.
+
+    This table allows us to store really large tensors separately from
+    the Tensors table. That way they don't slow down its scanning speed
+    and can be fetched in parallel.
+
+    Fields:
+      rowid: Must be same as corresponding Tensors table row.
+      tensor: A binary representation of this tensor, using the encoding
+          specified in the corresponding Tensors table row.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS BigTensors (
+          rowid INTEGER PRIMARY KEY,
+          tensor BLOB NOT NULL
+        )
+      ''')
+
+  def create_plugins_table(self):
+    """Creates the Plugins table.
+
+    This table exists to assign arbitrary IDs to TBPlugin names. These
+    IDs are handed out in monotonically increasing order, but that's OK,
+    because the size of this table will be extremely small. These IDs
+    will not be the same across TensorBoard installs.
+
+    It is assumed that once an ID is mapped to a name, that the mapping
+    will never change.
+
+    Fields:
+      plugin_id: Arbitrary 16-bit integer.
```

Comment:
Fine with me. If people are dynamically generating more than 2<sup>16</sup> plugins then we can cross that bridge when we come to it.